source/common/upstream/load_balancer_impl.cc

```diff
@@ -12,48 +12,88 @@
 
 namespace Envoy {
 namespace Upstream {
-namespace {
 
-const HostSet* bestAvailable(const PrioritySet* priority_set) {
-  if (priority_set == nullptr) {
-    return nullptr;
-  }
-  // This is used for LoadBalancerBase priority_set_ and local_priority_set_
-  // which are guaranteed to have at least one host set.
-  ASSERT(!priority_set->hostSetsPerPriority().empty());
+static const std::string RuntimeZoneEnabled = "upstream.zone_routing.enabled";
+static const std::string RuntimeMinClusterSize = "upstream.zone_routing.min_cluster_size";
+static const std::string RuntimePanicThreshold = "upstream.healthy_panic_threshold";
 
-  for (auto& host_set : priority_set->hostSetsPerPriority()) {
-    if (!host_set->healthyHosts().empty()) {
-      return host_set.get();
+uint32_t LoadBalancerBase::choosePriority(uint64_t hash,
+                                          const std::vector<uint32_t>& per_priority_load) {
+  hash = hash % 100 + 1; // 1-100
+  uint32_t aggregate_percentage_load = 0;
+  // As with tryChooseLocalLocalityHosts, this can be refactored for efficiency
+  // but O(N) is good enough for now given the expected number of priorities is
+  // small.
+  for (size_t priority = 0; priority < per_priority_load.size(); ++priority) {
+    aggregate_percentage_load += per_priority_load[priority];
+    if (hash <= aggregate_percentage_load) {
+      return priority;
     }
   }
-  return priority_set->hostSetsPerPriority()[0].get();
+  // The percentages should always add up to 100 but we have to have a return for the compiler.
+  NOT_REACHED;
+  return 0;
```

Comment:
nit: I don't think the compiler will require the return if you have NOT_REACHED.